File: CFOnlineShop/CFOnlineShop/ViewController/CFHomePageController/CFDetailInfoController.m
//
//  CFDetailInfoController.m
//  CFOnlineShop
//
//  Created by chenfeng on 2018/7/19.
//  Copyright © 2018年 chenfeng. All rights reserved.
//

#import "CFDetailInfoController.h"
#import "LPSemiModalView.h"
#import "CFSegmentedControl.h"
#import "CFDetailViewController.h"
#import "CFActivityController.h"
#import "CFOthersController.h"


@interface CFDetailInfoController ()<CFSegmentedControlDataSource,CFSegmentedControlDelegate,UIScrollViewDelegate>

//动画缩放视图
@property (nonatomic, strong) LPSemiModalView *narrowedModalView;

@property (nonatomic, strong) CFSegmentedControl *segmentedControl;

//能左右滑动的scrollview
@property (nonatomic, strong) UIScrollView *bgScrollView;

@property (nonatomic, strong) CFDetailViewController *detailViewController;

@property (nonatomic, strong) NSArray *segmentTitles;

@end

@implementation CFDetailInfoController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    _segmentTitles = @[@"详情",@"活动",@"其他"];
    
    self.navigationBgView.backgroundColor = kWhiteColor;
    self.navigationBgView.alpha = 0;
    [self setSegmentedControl];
    [self showLeftBackButton];
    [self setBgScrollview];
    [self setShadowAnimationView];
    [self setDetailView];
}

- (void)setBgScrollview
{
    _bgScrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, Main_Screen_Height)];
    _bgScrollView.delegate = self;
    _bgScrollView.showsVerticalScrollIndicator = NO;
    _bgScrollView.showsHorizontalScrollIndicator = NO;
    _bgScrollView.contentSize = CGSizeMake(Main_Screen_Width * [_segmentTitles count], Main_Screen_Height);
    _bgScrollView.pagingEnabled = YES;
    _bgScrollView.bounces = NO;
    [self.view addSubview:_bgScrollView];
}

- (void)setSegmentedControl
{
    //注意宽度要留够，不然title显示不完，title宽度是计算出来的。代码并不复杂，可以根据需要去内部进行修改
    _segmentedControl = [[CFSegmentedControl alloc] initWithFrame:CGRectMake(Main_Screen_Width/2 - (60 * [_segmentTitles count])/2, TopHeight - 40, 60 * [_segmentTitles count], 40)];
    _segmentedControl.delegate = self;
    _segmentedControl.dataSource = self;
    _segmentedControl.alpha = 0;
    [self.navigationView addSubview:_segmentedControl];
}

- (void)setShadowAnimationView
{
    _narrowedModalView = [[LPSemiModalView alloc] initWithSize:CGSizeMake(Main_Screen_Width, 300) andBaseViewController:self.navigationController];
    _narrowedModalView.contentView.backgroundColor = kWhiteColor;
    
    //显示内容在contentView上添加
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(_narrowedModalView.contentView.mj_w/2 - 50, 100, 100, 20)];
    label.textColor = KDarkTextColor;
    label.textAlignment = NSTextAlignmentCenter;
    label.font = SYSTEMFONT(16);
    label.text = @"暂无内容";
    [_narrowedModalView.contentView addSubview:label];
}

- (void)setDetailView
{
    //这里为了避免该控制器耦合性高的问题，所以使用addChildViewController的形式，来添加视图
    //详情
    _detailViewController = [[CFDetailViewController alloc] init];
    _detailViewController.image = _image;
    [self addChildViewController:_detailViewController];
    [_detailViewController didMoveToParentViewController:self];
    [_detailViewController.view setFrame:CGRectMake(0, 0, Main_Screen_Width, Main_Screen_Height)];
    //注意视图要添加到_bgScrollView上
    [_bgScrollView addSubview:_detailViewController.view];
    WeakSelf(self);
    [_detailViewController setScrollViewDidScroll:^(UIScrollView *scrollView) {
        if (scrollView == weakself.detailViewController.tableView){
            
            if (scrollView.mj_offsetY > 0 && scrollView.mj_offsetY < 60) {
                
                weakself.navigationBgView.alpha = 1 * (scrollView.mj_offsetY / 60.f);
                weakself.segmentedControl.alpha = weakself.navigationBgView.alpha;
            }
            else if (scrollView.mj_offsetY <= 0)
            {
                weakself.navigationBgView.alpha = 0;
                weakself.segmentedControl.alpha = weakself.navigationBgView.alpha;
            }
            else if (scrollView.mj_offsetY > 40)
            {
                if (weakself.navigationBgView.alpha != 1) {
                    weakself.navigationBgView.alpha = 1;
                    weakself.segmentedControl.alpha = weakself.navigationBgView.alpha;
                }
            }
            
        }
    }];
    [_detailViewController setAddActionWithBlock:^{
        [weakself.narrowedModalView open];
    }];
    
    //活动
    CFActivityController *activityController = [[CFActivityController alloc] init];
    [self addChildViewController:activityController];
    [activityController didMoveToParentViewController:self];
    [activityController.view setFrame:CGRectMake(Main_Screen_Width, 0, Main_Screen_Width, Main_Screen_Height)];
    [_bgScrollView addSubview:activityController.view];
    
    //其他
    CFOthersController *othersController = [[CFOthersController alloc] init];
    [self addChildViewController:othersController];
    [othersController didMoveToParentViewController:self];
    [othersController.view setFrame:CGRectMake(Main_Screen_Width*2, 0, Main_Screen_Width, Main_Screen_Height)];
    [_bgScrollView addSubview:othersController.view];
}

#pragma mark -- UIScrollViewDelegate 用于控制头部视图滑动的视差效果

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    if (scrollView == _bgScrollView) {
        
        NSInteger index = scrollView.mj_offsetX/Main_Screen_Width;
        
        [_segmentedControl didSelectIndex:index];
        WeakSelf(self);
        if (index == 0) {
            if (_detailViewController.tableView.mj_offsetY >= 0 && _detailViewController.tableView.mj_offsetY < 60) {
                
                [UIView animateWithDuration:0.25 animations:^{
                    weakself.navigationBgView.alpha = 1 * (weakself.detailViewController.tableView.mj_offsetY / 60.f);
                    weakself.segmentedControl.alpha = weakself.navigationBgView.alpha;
                }];
            }
        }
        else if (index >= 1)
        {
            [UIView animateWithDuration:0.25 animations:^{
                weakself.navigationBgView.alpha = 1;
                weakself.segmentedControl.alpha = weakself.navigationBgView.alpha;
            }];
        }
    }
}

#pragma mark -- SegmentedControlDelegate & datasource

- (NSArray *)getSegmentedControlTitles
{
    return _segmentTitles;
}

- (void)control:(CFSegmentedControl *)control didSelectAtIndex:(NSInteger)index
{
    [_bgScrollView setContentOffset:CGPointMake(Main_Screen_Width * index, 0) animated:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SwiftOnlineShopDemo/SwiftOnlineShopDemo/ViewController/CFHomePageController/CFDetailInfoController.swift
//
//  CFDetailInfoController.swift
//  SwiftOnlineShopDemo
//
//  Created by chenfeng on 2019/4/26.
//  Copyright © 2019年 chenfeng. All rights reserved.
//

import UIKit

class CFDetailInfoController: CFBaseController,CFSegmentedControlDataSource,CFSegmentedControlDelegate,UIScrollViewDelegate {
    
    var narrowedModalView: LPSemiModalView!//透视效果动画视图
    var segmentedControl: CFSegmentedControl!
    var bgScrollView: UIScrollView!
    var segmentTitles: Array<String>!
    var detailViewController: CFDetailViewController!
    

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        //self.modalPresentationStyle = .custom
        segmentTitles = ["详情","活动","其他"]
        navigationBgView.backgroundColor = UIColor.white;
        navigationBgView.alpha = 0;
        setSegmentedControl()
        showLeftBackButton()
        setBgScrollview();
        setShadowAnimationView();
        setDetailView();
    }
    
    func setSegmentedControl() {
        
        //注意宽度要留够，不然title显示不完，title宽度是计算出来的。代码并不复杂，可以根据需要去内部进行修改
        segmentedControl = CFSegmentedControl.init(frame: CGRect(x: Int(SCREEN_WIDTH/2) - (60 * segmentTitles.count)/2, y: Int(STATUS_AND_NAV_BAR_HEIGHT - 40), width: 60 * segmentTitles.count, height: 40))
        segmentedControl.delegate = self
        segmentedControl.dataSource = self
        segmentedControl.alpha = 0
        navigationView.addSubview(segmentedControl)
        
    }
    
    func setBgScrollview() {
        
        bgScrollView = UIScrollView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_HEIGHT))
        bgScrollView.delegate = self
        bgScrollView.showsVerticalScrollIndicator = false
        bgScrollView.showsHorizontalScrollIndicator = false
        bgScrollView.contentSize = CGSize(width: SCREEN_WIDTH * CGFloat(segmentTitles.count), height: SCREEN_HEIGHT)
        bgScrollView.isPagingEnabled = true
        bgScrollView.bounces = false
        self.view.addSubview(bgScrollView)
        //去掉顶部偏移
        if #available(iOS 11.0, *)
        {
            bgScrollView.contentInsetAdjustmentBehavior = .never;
        }
    }
    
    func setShadowAnimationView() {

        narrowedModalView = LPSemiModalView.init(size: CGSize(width: SCREEN_WIDTH, height: 300), andBaseViewController: navigationController)
        narrowedModalView.contentView.backgroundColor = UIColor.white
        
        let label = UILabel.init(frame: CGRect(x: narrowedModalView.contentView.mj_w/2 - 50, y: 100, width: 100, height: 20))
        label.textColor = UIColor.black
        label.textAlignment = .center
        label.font = UIFont.systemFont(ofSize: 16)
        label.text = "暂无内容"
        narrowedModalView.contentView.addSubview(label)
    
    }
    
    func setDetailView() {
        
        //这里为了避免该控制器耦合性高的问题，所以使用addChildViewController的形式，来添加视图
        //详情
        
        detailViewController = CFDetailViewController()
        self.addChild(detailViewController)
        detailViewController.didMove(toParent: self)
        detailViewController.view.frame = CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_HEIGHT)
        bgScrollView.addSubview(detailViewController.view)
        
        //swift闭包循环引用的问题：unowned表示确认当前闭包self不会提前释放；weak表示当前闭包self有可能会出现提前释放的情况，详细解释请查看官方文档
        detailViewController.scrollViewDidScroll_Closure = {[unowned self] (scrollView:UIScrollView)  in
            if scrollView == self.detailViewController.tableView{
                
                if scrollView.mj_offsetY > 0 && scrollView.mj_offsetY < 60 {
                    
                    self.navigationBgView.alpha = 1 * (scrollView.mj_offsetY / 60)
                    self.segmentedControl.alpha = self.navigationBgView.alpha
                }
                else if scrollView.mj_offsetY <= 0
                {
                    self.navigationBgView.alpha = 0
                    self.segmentedControl.alpha = self.navigationBgView.alpha
                }
                else if scrollView.mj_offsetY > 40
                {
                    if self.navigationBgView.alpha != 1 {
                        self.navigationBgView.alpha = 1
                        self.segmentedControl.alpha = self.navigationBgView.alpha
                    }
                }
                
            }
            
        }
        
        detailViewController.addActionWithBlock = {[unowned self] in
            self.narrowedModalView.open()
        }
        
        let activityController = CFActivityController()
        self.addChild(activityController)
        activityController.didMove(toParent: self)
        activityController.view.frame = CGRect(x: SCREEN_WIDTH, y: 140, width: SCREEN_WIDTH, height: SCREEN_HEIGHT - 300)
        bgScrollView.addSubview(activityController.view)
        
        let othersController = CFOthersController()
        self.addChild(othersController)
        othersController.didMove(toParent: self)
        othersController.view.frame = CGRect(x: SCREEN_WIDTH * 2, y: 0, width: SCREEN_WIDTH, height: SCREEN_HEIGHT)
        bgScrollView.addSubview(othersController.view)
        
    }
    
    // MARK: - SegmentedControl Method
    func getSegmentedControlTitles() -> Array<Any> {
        return segmentTitles
    }
    
    func control(_ control: CFSegmentedControl, didSelectAt Index: Int) {
        bgScrollView.setContentOffset(CGPoint(x: SCREEN_WIDTH * CGFloat(Index), y: 0), animated: true)
    }
    
    // MARK: - UIScrollView Method
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        if scrollView == bgScrollView {
            
            let index = scrollView.mj_offsetX/SCREEN_WIDTH;
            
            segmentedControl.didSelect(index: Int(index));
            if index == 0 {
                if detailViewController.tableView.mj_offsetY >= 0 && detailViewController.tableView.mj_offsetY < 60 {
                    UIView.animate(withDuration: 0.25) {
                        self.navigationBgView.alpha = 1 * (self.detailViewController.tableView.mj_offsetY / 60);
                        self.segmentedControl.alpha = self.navigationBgView.alpha;
                    }
                }
            }
            else if index >= 1
            {
                UIView.animate(withDuration: 0.25) {
                    self.navigationBgView.alpha = 1;
                    self.segmentedControl.alpha = self.navigationBgView.alpha;
                }
            }
        }
    }

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
